### ISSUE TITLE
CUDA device architecture

### ISSUE THREAD

**Issue by davydden** [spack/spack#6005]
Sometimes one needs to be explicit and provide `-arch=sm_60` or `-arch=sm_35` flags when building with CUDA. I don't know enough about CUDA but maybe Spack should know this information so that if CUDA is enabled, you can get something like `spec.cuda_arch` to return `sm_60` or alike. 

In meantime, one could probably encode this into a variant:
```
variant(
        'cuda',
        default=None,
        description='CUDA architecture',
        values=('sm_70', 'sm_62', 'sm_61', 'sm_60', ..., 'sm_35',...),
        multi=False
    )
```

But those values should also be consistent with CUDA version, from http://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
```
Supported on CUDA 7 and later
SM20 – Older cards such as GeForce GT630
SM30 – Kepler architecture (generic – Tesla K40/K80)
Adds support for unified memory programming
SM35 – More specific Tesla K40
Adds support for dynamic parallelism. Shows no real benefit over SM30 in my experience.
SM37 – More specific Tesla K80
Adds a few more registers. Shows no real benefit over SM30 in my experience
SM50 – Tesla/Quadro M series
SM52 – Quadro M6000 , GTX 980/Titan
SM53 – Tegra TX1 / Tegra X1
Supported on CUDA 8 and later
SM60 – GP100/Pascal P100 – DGX-1 (Generic Pascal)
SM61 – GTX 1080, 1070, 1060, Titan Xp, Tesla P40, Tesla P4
SM62 – Probably Drive-PX2
Supported on CUDA 9 and later
SM70 – Tesla V100
```

**Comment by tgamblin** [comment]
I *think* these should probably be variants (a variant?) on the CUDA package, with the default value potentially autodetected for the system (unless the user forces a build on another arch).

Does that sound reasonable?

**Comment by davydden** [comment]
> Does that sound reasonable?

yes. I am just not sure about autodetection. I though values of this variant should be limited to `None` (probably default) or specific architecture (i.e. `sm20`). If there is `auto` then this `auto` can potentially be any of those specific ones. So we get to a minor inconsistency where `cuda=auto` and `cuda=sm30` would be two different builds in Spack which, however, may be exactly the same.

The reason for `None` to be default is that CUDA, AFAIK, is very sensitive to used compilers and their version. Say CUDA8 does not support anything above gcc 5.4.0.

What do you say if we try to accumulate all those dependencies in an auxiliary package

```
class CUDAPackage(Package):
     variant(
        'cuda',
        default=None,
        description='CUDA architecture',
        values=('sm_70', 'sm_62', 'sm_61', 'sm_60', ..., 'sm_35',...),
        multi=False
    )

    depends_on('cuda@8:', when='cuda=sm_60')
    depends_on('cuda@8:', when='cuda=sm_61')
    depends_on('cuda@8:', when='cuda=sm_62')
    depends_on('cuda@9:', when='cuda=sm_70')

    conflicts(when='%gcc@6:^cuda@8:')
    // etc
```
to be used as
```
class Dealii(CMakePackage, CUDAPackage):
   // standard package
```
(or whatever is Python interface to derive from two classes)

I will play around with these ideas a couple of weeks when my colleague who uses `cuda` is back to the office so that we can test it.

**Comment by svenevs** [comment]
What do you intend to achieve by going down this route / how is inheriting from `CUDAPackage` going to affect the build environment?  It's generally very rare to find projects that expose this level of code generation control.

1. `cuda@9` and `sm_20` should `conflict`, they have officially (after years of promising xD) made this architecture obsolete.
2. If you aren't familiar, a brief overview of how it works.  Suppose you specifically compile for `sm_30,sm_35,sm_50`.  PTXAS instructions will be generated in the resultant binary for each different architecture (so it is compiled three times here).  At runtime, the branch most relevant to the device executing the code is selected.  So if you were running the code on an `sm_30` device, it would choose the `sm_30` branch.  If running an `sm_50` device, the `sm_50` branch.  Lastly, they are (to my knowledge) guaranteed to be backward compatible.  If the device running the code is `sm_60`, since no specific `sm_60` branch is compiled, the `sm_50` branch will be selected.
3. `sm_35` in particular is a very special one that should be preferred to ON where possible.  This was the introduction of so-called "dynamic parallelism", or allowing kernels to launch kernels.

The reason it gets complicated is because there are differences between the virtual and actual architecture.  With NVCC, it might look something like

```cmake
# Current arch's: http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#gpu-feature-list
set(BUILD_CUDA_ARCH "30;32;35;50;52;53;60;61;62;70" CACHE STRING
    "CUDA Architectures to build, default: 30;32;35;50;52;53;60;61;62;70")

# Add the compilation trajectories for different architectures.
# See: http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-examples
# I have no specific need for complicated build trajectories.
foreach(arch ${BUILD_CUDA_ARCH})
  # virtually real
  list(APPEND CUDA_NVCC_FLAGS "--generate-code arch=compute_${arch},code=sm_${arch}")
  # virtually virtual
  list(APPEND CUDA_NVCC_FLAGS "--generate-code arch=compute_${arch},code=compute_${arch}")
endforeach()
```

On the other hand, you can also use `llvm@3.9:`, where the virtual architectures are not supported.  You can do something like this:

```cmake
# See: http://llvm.org/docs/CompileCudaWithLLVM.html#compiling-cuda-code
foreach(arch ${BUILD_CUDA_ARCH})
  set(CLANG_CUDA_CXX_FLAGS "${CLANG_CUDA_CXX_FLAGS} --cuda-gpu-arch=sm_${arch}")
endforeach()
```

Then you might `set_source_files_properties` on specific cuda headers / source files with the accumulated flags.

Basically, I've never seen a consistent way of specifying architectures explicitly.  If you look at the (insane) `Makefile`s provided in the samples, they tend to just add `--generate-code arch=compute_X,code=sm_X` for everything (unless the demo explicitly needs a later version).

**Comment by davydden** [comment]
thanks for your descriptive comment @svenevs, I am indeed not much familiar with CUDA and thereby your comment is very helpful 👍 

> how is inheriting from CUDAPackage going to affect the build environment

my intention is to improve a rather limited 
```
variant('cuda', default=False,
            description='Build with CUDA')
```


So I though that `CUDAPackage` would NOT set-up any build environment and such, but only contain:

* a single multi-valued variant with architectures
* all the conflicts (architecture vs CUDA version / CUDA version vs compiler version) 
* and perhaps a single function to translate architecture into flags, i.e. 
```
--generate-code arch=compute_${arch},code=sm_${arch}
```
or 
```
--cuda-gpu-arch=sm_${arch}
```

Then it would be up to other packages which derive from `CUDAPackage` to use this in their build system (CMake/Autotools/etc).

The intention would be to have a single place to manage those CUDA specific constraints/options so that other packages do not have to re-implement them.

**Comment by svenevs** [comment]
I see.  So enforcing the constraints on the spack side should be doable, I'll look later tonight to see if I can find the official release page archives, I think they're still floating around.

Initial thought on implementation: multiple inheritance on `CudaPackage` is a good failsafe, but it would probably also be good to blacklist `variant('cuda')` (if possible) and inform the developer that they instead need to inherit from `CudaPackage`.

### Main Problem

While enforcing the soft constraints on e.g. `^cuda@9+sm_20` as a conflict in Spack would be relatively straightforward, this basically requires every package using `depends_on('cuda')` in a position where to satisfy the Spack constraint, the build system must be patched to actually make it so this is true.  In most CMake packages, it could be as simple as hacking in the above NVCC CMake code into their build system.  For Autotools and other stuff, if they are careful to read `NVCC_FLAGS` or something, this would also be potentially easy to patch.  It's more just the idea that now every build system would have to be inspected to actually make the Spack constraint true.

### Autodetection

Spack could potentially automate this.  Below is a rough prototype, it produces a single architecture per line or the string `[X] No CUDA devices found.` to signal a problem.

```cpp
// nvcc -o spack_cuda_compute_capabilities spack_cuda_compute_capabilities.cpp
#include <cuda_runtime.h>
#include <iostream>

int main(const int /*argc*/, const char ** /*argv*/) {
    int deviceCount = 0;
    cudaGetDeviceCount(&deviceCount);

    if(deviceCount == 0) {
        std::cout << "[X] No CUDA devices found." << std::endl;
    }

    for(int i = 0; i < deviceCount; ++i) {
        cudaDeviceProp dp;
        cudaGetDeviceProperties(&dp, i);
        std::cout << "sm_" << dp.major << dp.minor << std::endl;
    }
}
```

This is not nearly as useful as it may initially seem.  For a desktop / laptop, it works just fine.  Where clusters are concerned, users would need to manually submit jobs to determine what the compute architectures are if they don't know them.  E.g. if their Spack installation is on the head node, but the worker nodes are the ones with the actual GPUs attached.

### -WMaybe-Unnecessary

I think `nvcc` may actually default to building all targets.  I'm trying to find docs on that, but I think I'm going to ask on the CUDA forums if people have any suggestions for this discussion.

```console
$ nvcc -o spack_cuda_compute_capabilities spack_cuda_compute_capabilities.cpp
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
```

This was with CUDA 8 on OSX, I'll check Linux when I'm back at that box.

### Related Thought

The only other thing that comes to mind, specifically where LLVM is concerned, is some kind of `provides` mechanism.  Personally, though, I think that just not using LLVM to compile any CUDA is a better decision.  It may become a logistical nightmare, since although LLVM can compile, it still needs the actual CUDA toolkit.  There are differences in flags, as well as static vs dynamic linkage depends on the operating system.  Essentially, if a build system has inside of it the ability to either use NVCC or use LLVM, assume they have an `option` available to trigger this.  I doubt any packages in Spack have this ability, this is all generally very recent.

**Comment by davydden** [comment]
@svenevs i created a WIP PR to illustrate the idea https://github.com/LLNL/spack/pull/6070

**Comment by ax3l** [comment]
@davydden 

actually, a variant `cuda` would be `multi=True` :) it's very common to build multiple PTX and SASS artifacts into a single fat binary. See e.g. [alpaka](https://github.com/ComputationalRadiationPhysics/alpaka/blob/32b3dcf6ad032f60d7a56768b8f1221aefa6f80b/alpakaConfig.cmake#L359-L363).

Here is the official list of SMs vs. devices & generations:
  https://developer.nvidia.com/cuda-gpus

note that they are forward compatible.

**Comment by ax3l** [comment]
linking to my issue in #2379 :)

**Comment by ax3l** [comment]
> AFAIK, is very sensitive to used compilers and their version

yes, if you want the full compatebility matrix see [my gist](https://gist.github.com/ax3l/9489132).

I also added a spack-conflict-representation in our [PIConGPU builds script](https://github.com/ComputationalRadiationPhysics/spack-repo/blob/9c05bba8158a31f3f7f233be634345c4cb37fd54/packages/picongpu/package.py#L85-L94). Please be aware that those conflicts are a `nvcc` constrain, *not* a CUDA constrain (e.g. clang 3.9/4.0+ are excellent CUDA device compilers and do not need `nvcc`).

**Comment by davydden** [comment]
@ax3l thanks for the links! Those conflicts and alike are something I try to initiate in that other WIP PR.

**Comment by ax3l** [comment]
@svenevs actually the clang world is not too complicated for CUDA device compilation. They need the [CUDA SDK](https://bugs.llvm.org/show_bug.cgi?id=26966#c6) and projects usually just set the a cuda device compiler via a CMake var to `clang` instead of `nvcc`.

The problem now is, that such a var [is not yet standardized in CMake](https://twitter.com/robertjmaynard/status/916318744133201920) although there technically is a `CUDA_NVCC_EXECUTABLE` in CMake's `FindCUDA.cmake`, but e.g. cuda source files [are passed slightly differently to clang and nvcc](https://github.com/ComputationalRadiationPhysics/alpaka/blob/32b3dcf6ad032f60d7a56768b8f1221aefa6f80b/cmake/addExecutable.cmake#L29-L52). But those interface discrepancies are as neat as with "normal" host compilers :)

**Comment by svenevs** [comment]
@ax3l thanks for all of the great information on this!

The concern about LLVM being able to be supplied as far as Spack is concerned is more of the logistics behind enabling it.

**Does the package detect clang 3.9 and automatically switch?**  If so, Spack has little control.

**Does the package assume NVCC?** If so, Spack would have to patch to change things (e.g., `CUDA_NVCC_EXECUTABLE` getting changed (?), or more heavyweight patches).  The fast math flags are different, specifying architectures is different, etc.  It just seems like it would be really challenging to pull off.

**Comment by ax3l** [comment]
It's actually similar to possible future compiler abstractions, in terms of having a "virtual" package that provides C++11 or similar. In such a case, one would have a compiler that can compile CUDA C++ and clang is just an other possibility for it.

We currently think to address this as a variant of the package simply (`cudacxx`).
